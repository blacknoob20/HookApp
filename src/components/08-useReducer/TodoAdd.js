import React from 'react';
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({ handleAddTodo }) => {
    const [{ description }, handleInputChange, reset] = useForm({ description: '' });

    const handleSubmit = (e) => {
        e.preventDefault();

        if (description.trim().length < 1) return;

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false,
        };

        handleAddTodo(newTodo);
        reset();
    };

    return (
        <>
            <h4>Agregar TODO</h4>
            <hr />
            <form action="">
                <input type="text" name="description" placeholder="Descripcion" autoComplete="off" className="form-control" onChange={handleInputChange} value={description} />
                <button onClick={handleSubmit} className="btn btn-outline-primary btn-lg btn-block mt-2">Agregar</button>
            </form>
        </>
    )
}
