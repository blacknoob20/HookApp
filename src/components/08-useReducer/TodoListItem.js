import React from 'react'

export const TodoListItem = (todo, i, handleToggle, handleDelete) => {
    return (
        <li key={todo.id} className="list-group-item">
            <span onClick={e => handleToggle(todo.id)} className={`${todo.done && 'complete'}`}>{i + 1}. {todo.desc}</span>
            <button onClick={e => handleDelete(todo.id)} className="btn btn-danger mx-3">Borrar</button>
        </li>
    )
}
