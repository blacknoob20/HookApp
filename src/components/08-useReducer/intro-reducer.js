const initState = [{
    id: 1,
    todo: 'Comprar leche',
    done: false,
}];

const todoReducer = (state = initState, action) => {
    if (action?.type === 'agregar') return [...state, action.payload];

    return state;
}

let todos = todoReducer();
const newTodo = {
    id: 2,
    todo: 'Comprar queso',
    done: false,
};

// La accion que se envia al reducer como norma
// debe tener el atributo type
// Opcional
// el atributo payload, que es donde se envia la data
const action = {
    type: 'agregar',
    payload: newTodo

}

todos = todoReducer(todos, action);

console.log(todos);