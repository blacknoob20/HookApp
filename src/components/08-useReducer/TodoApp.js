import React, { useEffect, useReducer } from 'react';
import { todoReducer } from './todoReducer';

import './style.css';
import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';

const init = () => JSON.parse(localStorage.getItem('todos')) || [];

export const TodoApp = () => {
    // El dispatch es una funcion que toma una accion y la setea al reducer correspondiente
    const [todos, dispatch] = useReducer(todoReducer, [], init);

    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos));
    }, [todos]);

    const handleDelete = todoId => {
        console.log(todoId);
        const action = { type: 'del', payload: todoId };

        dispatch(action);
    };

    const handleToggle = todoId => dispatch({ type: 'tog', payload: todoId });

    const handleAddTodo = newTodo => dispatch({ type: 'add', payload: newTodo });

    return (
        <>
            <h1>TodoApp ({todos.length})</h1>
            <hr />
            <div className="container">
                <div className="row">
                    <div className="col">
                        <TodoList todos={todos} handleToggle={handleToggle} handleDelete={handleDelete} />
                    </div>
                    <div className="col">
                        <TodoAdd handleAddTodo={handleAddTodo} />
                    </div>
                </div>
            </div>
        </>

    )
}
