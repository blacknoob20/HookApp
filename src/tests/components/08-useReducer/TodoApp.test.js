import { act } from '@testing-library/react';
import { mount, shallow } from 'enzyme';
import { TodoApp } from '../../../components/08-useReducer/TodoApp';
import { demosTodos } from '../../fixtures/demoTodos';

describe('Pruebas en <TodoApp/>', () => {
    const wrapper = shallow(<TodoApp />);

    Storage.prototype.setItem = jest.fn();

    test('Debe mostrarse correctamente.', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('Debe de agregar un TODO', () => {
        const wrapper = mount(<TodoApp />);
        act(() => {
            wrapper.find('TodoAdd').prop('handleAddTodo')(demosTodos[0]);
        });

        expect(wrapper.find('h1').text().trim()).toBe('TodoApp (1)');
        expect(localStorage.setItem).toHaveBeenCalledTimes(2);
    });

    test('Debe eliminar un TODO.', () => {
        wrapper.find('TodoAdd').prop('handleAddTodo')(demosTodos[0]);
        wrapper.find('TodoList').prop('handleDelete')(demosTodos[0].id);

        expect(wrapper.find('h1').text().trim()).toBe('TodoApp (0)');
    });
    


});
