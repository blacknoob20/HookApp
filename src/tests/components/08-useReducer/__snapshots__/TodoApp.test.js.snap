// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoApp/> Debe mostrarse correctamente. 1`] = `
<Fragment>
  <h1>
    TodoApp (
    0
    )
  </h1>
  <hr />
  <div
    className="container"
  >
    <div
      className="row"
    >
      <div
        className="col"
      >
        <TodoList
          handleDelete={[Function]}
          handleToggle={[Function]}
          todos={Array []}
        />
      </div>
      <div
        className="col"
      >
        <TodoAdd
          handleAddTodo={[Function]}
        />
      </div>
    </div>
  </div>
</Fragment>
`;
